src/compiler/Fork.js

```diff
@@ -0,0 +1,121 @@
+/**
+ * Copyright 2017-present, Callstack.
+ * All rights reserved.
+ * 
+ * @flow
+ */
+
+import type { Platform } from '../types';
+
+const fs = require('fs');
+const EventEmitter = require('events');
+const Events = require('./events');
+const createForkProcess = require('./createForkProcess');
+const createWebSocketServer = require('./createWebSocketServer');
+
+type ForkConstructorArgs = {
+  platform: Platform,
+  options: *,
+};
+
+const forks = {};
+let transportServer;
+
+/**
+ * Fork class is a abstraction over fork process. It handles creation of the process,
+ * communication over transport server (WebSocket server) and redirecting events from
+ * Webpack compiler in fork process to Fork consumer.
+ */
+module.exports = class Fork extends EventEmitter {
+  platform: Platform;
+  process: any;
+  socket: WebSocket;
+  enqueuedMessages: any[];
+  isProcessing: boolean;
+
+  constructor({ platform, options }: ForkConstructorArgs) {
+    super();
+
+    if (!transportServer) {
+      transportServer = createWebSocketServer();
+
+      // WebSocket connection is established after the Fork is created.
+      transportServer.on('connection', socket => {
+        const platformMatch = socket.upgradeReq.url.match(
+          /platform=(ios|android)/
+        );
+
+        if (!platformMatch) {
+          throw new Error('Incorrect platform');
+        }
+
+        forks[platformMatch[1]].setSocket(socket);
+      });
+    }
+
+    this.isProcessing = true;
+    this.enqueuedMessages = [];
+    this.platform = platform;
+    this.process = createForkProcess(
+      platform,
+      __dirname,
+      transportServer.options.server.address(),
+      options
+    );
+
+    forks[platform] = this;
+  }
+
+  setSocket(socket: WebSocket) {
+    this.socket = socket;
+
+    // Flush enqueued messages.
+    this.enqueuedMessages.forEach(({ type, ...payload }) =>
```

Comment:
Could use `.filter` instead of `.forEach` and just return `false` each time, but meh